python/lsst/summit/extras/plotting/focusSweep.py

```diff
@@ -0,0 +1,282 @@
+        for k in ["x", "y", "z", "u", "v", "age"]:
+            data[prefix + k] = np.nan
+
+    for row in data:
+        rIdx = [r.id for r in records].index(row["visit_id"])
+        record = records[rIdx]
+        for pid, prefix in zip(["MTHexapod:1", "MTHexapod:2"], ["cam_", "m2_"]):
+            try:
+                efdData = getMostRecentRowWithDataBefore(
+                    efdClient,
+                    "lsst.sal.MTHexapod.logevent_compensatedPosition",
+                    timeToLookBefore=record.timespan.begin,
+                    maxSearchNMinutes=3,
+                    where=lambda df: df["private_identity"] == pid,
+                )
+            except ValueError:
+                row[prefix + "age"] = np.nan
+                for k in ["x", "y", "z", "u", "v"]:
+                    row[prefix + k] = np.nan
+            else:
+                age = record.timespan.begin - efdTimestampToAstropy(efdData["private_efdStamp"])
+                row[prefix + "age"] = age.sec
+                for k in ["x", "y", "z", "u", "v"]:
+                    row[prefix + k] = efdData[k]
+    return data
+
+
+def inferSweepVariable(data):
+    """Heuristically determine which variable is being swept during a focus
+    sweep.
+
+    Parameters
+    ----------
+    data : astropy.table.Table
+        Table holding sweep hexapod motions and PSF measurements.
+
+    Returns
+    -------
+    varName : str | None
+        Name of the inferred active hexapod variable or None if inference
+        failed.
+    """
+    # Examine the ratio of RMS hexapod values to RMS hexapod residuals from a
+    # linear fit against seqNum.  If removing the linear term significantly
+    # reduces the RMS, that's a good sign this is the active variable.
+    stats = {}
+    for prefix in ["cam_", "m2_"]:
+        for k in ["x", "y", "z", "u", "v"]:
+            hexapodValue = data[prefix + k].value.astype(float)
+            seqNum = data["seqNum"]
+            coefs = np.polyfit(seqNum, hexapodValue, 1)
+            resids = np.polyval(coefs, seqNum) - hexapodValue
+            stdResids = np.nanstd(resids)
+            if stdResids == 0:
+                stats[prefix + k] = np.nan
+            else:
+                stats[prefix + k] = np.nanstd(hexapodValue) / stdResids
+    statMax = -np.inf
+    varName = None
+    for vName, stat in stats.items():
+        if stat > statMax:
+            varName = vName
+            statMax = stat
+    return varName
+
+
+def fitSweepParabola(data, varName):
+    fwhms = data["fwhm"]
+    e1s = data["e1"]
+    e2s = data["e2"]
+    xs = data[varName]
+    coefs, cov = np.polyfit(xs, fwhms, 2, cov=True)
+    a, b, c = coefs
+    vertex = -b / (2 * a)
+    resids = np.polyval(coefs, xs) - fwhms
+    rms = np.sqrt(np.mean(np.square(resids)))
+    extremum = np.polyval(coefs, vertex)
+
+    # WARNING!  Trusting ChatGPT with vertex uncertainty
+    # propagation for the moment.  Treat with extreme caution!
```

Comment:
Maybe add a TODO with a ticket number for this.

Reply:
Haha, will add. Again, this code is all Josh's, though you can obviously be forgiven for thinking this was all me 😄